cli/src/alluxio.org/cli/processes/common.go

```diff
@@ -0,0 +1,172 @@
+		if strings.Index(line, "#") != -1 {
+			line = line[:strings.Index(line, "#")]
+		}
+		line = strings.TrimSpace(line)
+		if line != "" {
+			mastersList = append(mastersList, line)
+		}
+	}
+	return mastersList, nil
+}
+
+func getWorkers() ([]string, error) {
+	workersDir := path.Join(env.Env.EnvVar.GetString(env.ConfAlluxioConfDir.EnvVar), "workers")
+	workersFile, err := os.Open(workersDir)
+	if err != nil {
+		log.Logger.Errorf("Error reading worker hostnames at %s", workersDir)
+		return nil, err
+	}
+
+	workersReader := bufio.NewReader(workersFile)
+	var workersList []string
+	lastLine := false
+	for !lastLine {
+		// read lines of the workers file
+		line, err := workersReader.ReadString('\n')
+		if err != nil {
+			if err == io.EOF {
+				lastLine = true
+			} else {
+				log.Logger.Errorf("Error parsing worker file at this line: %s", line)
+				return nil, err
+			}
+		}
+		// remove notes
+		if strings.Index(line, "#") != -1 {
+			line = line[:strings.Index(line, "#")]
+		}
+		line = strings.TrimSpace(line)
+		if line != "" {
+			workersList = append(workersList, line)
+		}
+	}
+	return workersList, nil
+}
+
+func getPrivateKey() (ssh.Signer, error) {
+	homePath, err := os.UserHomeDir()
+	if err != nil {
+		log.Logger.Errorf("User home directory not found at %s", homePath)
+		return nil, err
+	}
+	privateKey, err := os.ReadFile(path.Join(homePath, ".ssh", "id_rsa"))
+	if err != nil {
+		log.Logger.Errorf("Private key file not found at %s", path.Join(homePath, ".ssh", "id_rsa"))
+		return nil, err
+	}
+	parsedPrivateKey, err := ssh.ParsePrivateKey(privateKey)
+	if err != nil {
+		log.Logger.Errorf("Cannot parse public key at %s", path.Join(homePath, ".ssh", "id_rsa"))
+		return nil, err
+	}
+	return parsedPrivateKey, nil
+}
+
+func dialConnection(remoteAddress string, key ssh.Signer) (*ssh.Client, error) {
+	clientConfig := &ssh.ClientConfig{
+		// TODO: how to get user name? Like ${USER} in alluxio-common.sh
+		User: "root",
+		Auth: []ssh.AuthMethod{
+			ssh.PublicKeys(key),
+		},
+		Timeout:         5 * time.Second,
+		HostKeyCallback: ssh.InsecureIgnoreHostKey(),
+	}
+	// TODO: Some machines might have changed default SSH port. Get ssh port or remind users when get started.
+	dialAddr := fmt.Sprintf("%s:%d", remoteAddress, 22)
+	conn, err := ssh.Dial("tcp", dialAddr, clientConfig)
+	if err != nil {
+		log.Logger.Errorf("Dial failed to %s, error: %s", remoteAddress, err)
+	}
+	return conn, err
+}
+
+func closeConnection(remoteAddress string, conn *ssh.Client) error {
+	err := conn.Close()
+	if err != nil {
+		log.Logger.Infof("Connection to %s closed. Error: %s", remoteAddress, err)
+	} else {
+		log.Logger.Infof("Connection to %s closed.", remoteAddress)
+	}
+	return err
+}
+
+func runCommand(remoteAddress string, conn *ssh.Client, command string) error {
+	// create a session for each worker
+	session, err := conn.NewSession()
+	if err != nil {
+		log.Logger.Errorf("Cannot create session at %s", remoteAddress)
+		return err
+	}
+
```

Comment:
`defer conn.Close()` after opening session to ensure it is closed regardless of how the function exits. currently if there is an error in sesssion.Run(), close will never be called, causing a resource leak

Reply:
actually this is session.Close()